### PR TITLE
Add tags support to gateway backends

### PR DESCRIPTION
**Pull request by maswin**
## Description
The backend stored in Gateway today has only limited fields (external url, proxy-to, name, active/not-active).
But to make some sophisticated routing decision, having tags on them would help. 
For instance the version of the cluster, environment (prod/test), team name etc.
This PR does the following -

- Flyway V5 migrations for PostgreSQL, MySQL, and Oracle
- GatewayBackend DAO record and GatewayBackendDao updated to include
  tags in INSERT/UPDATE queries
- ProxyBackendConfiguration exposes tags as List<String>
- HaGatewayManager converts between comma-separated DB storage and
  List<String>, with tagsToString/tagsFromString helpers
- GatewayWebAppResource propagates tags in getAllBackends response
- UI: Tags column in the backends table rendered as tag chips;
  TagInput in the create/edit modal with space-or-enter to add tags
  and × to remove them; editTagsRef ensures tags survive any
  Form.TagInput closure/submission quirks
- Unit tests for tagsToString/tagsFromString edge cases and
  integration tests covering create, update, clear, and list

## Additional context and related issues
Tags in UI - 
Screenshot 2026-03-02 at 3 44 46 PM

Update box -
Screenshot 2026-03-02 at 3 44 58 PM

Tags updated - 
Screenshot 2026-03-02 at 3 45 18 PM

## Release notes
Add tags to gateway backends

( ) This is not user-visible or is docs only, and no release notes are required.
( X ) Release notes are required, with the following suggested text:

```markdown
Add support for attaching tags to backends
* [:warning: Breaking change:](#breaking) Add a `tags` column to the `gateway_backend` table. Deployments with `runMigrationsEnabled` set to `false` must apply `V5__add_tags_to_gateway_backend.sql` manually before upgrading, otherwise all backend create and update requests fail, even when no tags are used.
```

### CODE REVIEW

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManager.java [hunk 1]
```diff
@@ -150,6 +151,109 @@ void testGatewayManagerCacheExpire()
         assertThatThrownBy(() -> haGatewayManager.getBackendByName("new-etl1")).hasMessage("Failed to load backends from database to cache");
     }
 
+    @Test
+    void testAddBackendWithTags()
+    {
+        JdbcConnectionManager connectionManager = createTestingJdbcConnectionManager(dataStoreConfig());
+        HaGatewayManager haGatewayManager = new HaGatewayManager(connectionManager.getJdbi(), new RoutingConfiguration(), new DatabaseCacheConfiguration());
+
+        var backend = new ProxyBackendConfiguration();
```
Comment (vishalya) [inline gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManager.java:160]: [Nit: Avoid vars! ](https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-var)

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java [hunk 2]
```diff
@@ -16,13 +16,17 @@
...
+import java.util.List;
+
 public class ProxyBackendConfiguration
 {
     private boolean active = true;
     private String routingGroup = "adhoc";
     private String externalUrl;
     private String name;
     private String proxyTo;
+    private List<String> tags = Collections.emptyList();
```
Comment (vishalya) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java:29]: Use immutable list. `ImmutableList.of();` here and in the setter.

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+
+final class TestHaGatewayManagerTagHelpers
+{
+    // -----------------------------------------------------------------------
+    // tagsToString
+    // -----------------------------------------------------------------------
+
+    @Test
+    void tagsToStringReturnsNullForNullInput()
```
Comment (vishalya) [inline gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java:31]: prepend method name with `test` here and other places

File: docs/backend-tags.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+- **Team-based routing** — Tag backends with `team:bi-tools` or `team:data-eng`
+  and steer each team's queries to their dedicated cluster.
+- **Size-based routing** — Tag backends with `size:S`, `size:M`, or `size:XL`
+  and route large, resource-heavy queries to large clusters. Large queries can
+  hog resources for a long time, so routing them to appropriately sized clusters
+  reduces wait times for other queries.
+
+### Cluster upgrades and canary deployments
+
+- **Version tags** — Tag backends with `ver:476` to identify which Trino version
```
Comment (vishalya) [inline docs/backend-tags.md:74]: That is very useful!
Make sure to elaborate in the documentation.

File: gateway-ha/src/test/resources/add_backends_postgres.sql [hunk 5]
```diff
@@ -1,5 +1,5 @@
 INSERT INTO gateway_backend
-(name, routing_group, backend_url, external_url, active)
+(name, routing_group, backend_url, external_url, active, tags)
 VALUES
-('trino-1', 'adhoc', 'http://localhost:8081', 'http://localhost:8081', true),
-('trino-2', 'adhoc', 'http://localhost:8082', 'http://localhost:8082', true);
\ No newline at end of file
+('trino-1', 'adhoc', 'http://localhost:8081', 'http://localhost:8081', true, 'local,prod'),
```
Comment (vishalya) [inline gateway-ha/src/test/resources/add_backends_postgres.sql:4]: Is there input validation in place to prevent commas in tag values? what if we have a tag like `"version - 480,475"`

File: gateway-ha/src/main/java/io/trino/gateway/ha/resource/GatewayWebAppResource.java
Comment (vishalya) [inline gateway-ha/src/main/java/io/trino/gateway/ha/resource/GatewayWebAppResource.java:220]: If the payload doesn't send the tags, they will be cleared. But this code can inspect the payload and decide to use the sent one vs existing one.

**Review by vishalya — CHANGES_REQUESTED** [review]

File: gateway-ha/src/test/resources/add_backends_postgres.sql (hunk 5 above)
Comment (maswin) [reply gateway-ha/src/test/resources/add_backends_postgres.sql]: Good catch. I have added a validation such that tags cannot have "," in them. Also added a note in the docs to avoid comma and if they need something like `"version - 480,475"` to use `"version - 480","version - 475"`

File: docs/backend-tags.md (hunk 4 above)
Comment (maswin) [reply docs/backend-tags.md:83]: Done

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java (hunk 3 above)
Comment (maswin) [reply gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java]: Done

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java (hunk 2 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: Done

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManager.java (hunk 1 above)
Comment (maswin) [reply gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManager.java]: Done

File: gateway-ha/src/main/java/io/trino/gateway/ha/resource/GatewayWebAppResource.java
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/resource/GatewayWebAppResource.java:207]: This is true for every other field. Adding a special update logic for tags might make it inconsistent with other APIs. The UI always sends a full list of tags.

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManager.java (hunk 1 above)
Comment (ebyhr) [reply gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManager.java]: Here's the latest rule: 
```
Use var judiciously
Use var only when it improves readability. Prefer it when the type is obvious from the initializer, such as with new expressions, or when the explicit type is long or heavily generic and adds noise without improving clarity.

Avoid var when the inferred type is unclear, surprising, or important to understanding the code.
```

https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#use-var-judiciously

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        assertThat(tagsToString(List.of("prod"))).isEqualTo("prod");
+    }
+
+    @Test
+    void testTagsToStringJoinsMultipleTagsWithComma()
+    {
+        assertThat(tagsToString(List.of("prod", "us-east-1", "critical")))
+                .isEqualTo("prod,us-east-1,critical");
+    }
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java]: A test method per assertion is overkill. Please merge them into one. Same for tagsFromString.

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static io.trino.gateway.ha.router.HaGatewayManager.tagsFromString;
+import static io.trino.gateway.ha.router.HaGatewayManager.tagsToString;
+import static org.assertj.core.api.Assertions.assertThat;
+
+final class TestHaGatewayManagerTagHelpers
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java]: Remove this class, and move tests to TestHaGatewayManager.

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java [hunk 8]
```diff
@@ -86,4 +90,16 @@ public void setRoutingGroup(String routingGroup)
...
+    @JsonProperty
+    public List<String> getTags()
+    {
+        return this.tags;
+    }
+
+    @JsonSetter
+    public void setTags(List<String> tags)
+    {
+        this.tags = tags != null ? tags : ImmutableList.of();
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: Consider using `requireNonNullElse` method.

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java [hunk 9]
```diff
@@ -86,4 +90,16 @@ public void setRoutingGroup(String routingGroup)
     {
         this.routingGroup = routingGroup;
     }
+
+    @JsonProperty
+    public List<String> getTags()
+    {
+        return this.tags;
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]:
```suggestion
        return tags;
```

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java [hunk 10]
```diff
@@ -247,11 +252,28 @@ private static List<ProxyBackendConfiguration> upcast(List<GatewayBackend> gatew
...
             backendConfig.setName(model.name());
+            backendConfig.setTags(tagsFromString(model.tags()));
             proxyBackendConfigurations.add(backendConfig);
         }
         return proxyBackendConfigurations;
     }
 
+    static String tagsToString(List<String> tags)
+    {
+        if (tags == null || tags.isEmpty()) {
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:265]: Consider using `isNullOrEmpty` method. Same for tagsFromString.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java [hunk 11]
```diff
@@ -247,11 +252,28 @@ private static List<ProxyBackendConfiguration> upcast(List<GatewayBackend> gatew
...
+            return null;
+        }
+        return String.join(",", tags);
+    }
+
+    static List<String> tagsFromString(String tags)
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:272]: Add `@VisibleForTesting` annotation. However, these helper methods are pretty simple. There is no need to add tests, though.

Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:272]: Done

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java (hunk 10 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:265]: Done

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java (hunk 9 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: Every other get in the class has this.<variable_name>.
Why do we want to change it?

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java (hunk 8 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: Done

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java (hunk 7 above)
Comment (maswin) [reply gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java]: Done

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java (hunk 6 above)
Comment (maswin) [reply gateway-ha/src/test/java/io/trino/gateway/ha/router/TestHaGatewayManagerTagHelpers.java]: Done

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java (hunk 10 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:265]: `isNullOrEmpty` is not there for List (atleast with existing dependency), but its there for String and updated the method below with that.

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java (hunk 9 above)
Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: `this.` is redundant in the getter. Other places can be changed in a separate PR.

Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: done

**Comment by Chaho12** [comment]
Can you sqaush & rewrite commit message ?

**Comment by maswin** [comment]
> Can you sqaush & rewrite commit message ?

@Chaho12 - Done

**Comment by maswin** [comment]
@vishalya - can you review the changes?

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java [hunk 12]
```diff
@@ -229,6 +231,10 @@ private static void validateBackendConfiguration(ProxyBackendConfiguration backe
         checkArgument(backend.getProxyTo() != null, "Backend proxyTo URL cannot be null");
         checkArgument(backend.getRoutingGroup() != null, "Backend routing group cannot be null");
         checkArgument(backend.getExternalUrl() != null, "Backend external url cannot be null");
+        for (String tag : backend.getTags()) {
+            checkArgument(tag != null && !tag.isBlank(), "Backend tag cannot be null or blank");
+            checkArgument(!tag.contains(","), "Backend tag cannot contain ',' character: %s", tag);
```
Comment (Chaho12) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:236]: Read does trimResults, but write doesn't. Write/read asymmetry. We should update setTags so we fix at the source so that object is canonical eveywhere.

Comment (Chaho12) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:236]: no need for line 236, as comma is rejected anyway in line 268, as it uses `,` as storage delimiter.

File: webapp/src/components/cluster.tsx [hunk 13]
```diff
@@ -16,6 +16,13 @@ export function Cluster() {
   const [visibleForm, setVisibleForm] = useState(false);
   const [formApi, setFormApi] = useState<FormApi<any>>();
   const [form, setForm] = useState<BackendData>();
+  const [editTags, setEditTags] = useState<string[]>([]);
+  const editTagsRef = useRef<string[]>([]);
+
+  const updateEditTags = (tags: string[]) => {
+    editTagsRef.current = tags;
+    setEditTags(tags);
+  };
```
Comment (Chaho12) [inline webapp/src/components/cluster.tsx]: Isn't this redundant? As Form.TagInput already feeds value.tags (L233) into onSubmit (L171).

File: webapp/src/components/cluster.tsx [hunk 14]
```diff
@@ -45,6 +52,7 @@ export function Cluster() {
       <ButtonGroup size={'default'}>
         <Button onClick={() => {
           setForm(record)
+          updateEditTags(record.tags || [])
```
Comment (Chaho12) [inline webapp/src/components/cluster.tsx]: no need

File: webapp/src/components/cluster.tsx [hunk 15]
```diff
@@ -125,6 +145,7 @@ export function Cluster() {
               <ButtonGroup size={'default'}>
                 <Button onClick={() => {
                   setForm(undefined)
+                  updateEditTags([])
```
Comment (Chaho12) [inline webapp/src/components/cluster.tsx]: same here

File: webapp/src/components/cluster.tsx [hunk 16]
```diff
@@ -148,15 +169,16 @@ export function Cluster() {
           labelWidth={150}
           style={{ paddingRight: '20px' }}
           onSubmit={(values) => {
+            const payload = { ...values, tags: editTagsRef.current };
```
Comment (Chaho12) [inline webapp/src/components/cluster.tsx]: no need

File: webapp/src/components/cluster.tsx [hunk 17]
```diff
@@ -148,15 +169,16 @@ export function Cluster() {
...
             if (form === undefined) {
-              backendSaveApi(values)
+              backendSaveApi(payload)
```
Comment (Chaho12) [inline webapp/src/components/cluster.tsx]: keep original

File: webapp/src/components/cluster.tsx [hunk 18]
```diff
@@ -207,6 +229,14 @@ export function Cluster() {
             ]}
             initValue={form?.externalUrl}
           />
+          <Form.TagInput
+            field="tags"
+            label="Tags"
+            initValue={editTags}
```
Comment (Chaho12) [inline webapp/src/components/cluster.tsx]:
```suggestion
            initValue={form?.tags}
```

File: webapp/src/components/cluster.tsx [hunk 19]
```diff
@@ -207,6 +229,14 @@ export function Cluster() {
...
+            separator={[' ']}
+            onChange={(v) => updateEditTags(v as string[])}
```
Comment (Chaho12) [inline webapp/src/components/cluster.tsx]: no need

**Comment by mosabua** [comment]
@maswin can you address comments and work with @ebyhr and @Chaho12 ? I would like to get this feature in.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java (hunk 12 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:236]: The checkArgument is still needed - an un-rejected "a,b" tag would silently split into two tags on read rather than being rejected. This validation was also requested earlier in the review and is covered by tests.

File: webapp/src/components/cluster.tsx (hunk 15 above)
Comment (maswin) [reply webapp/src/components/cluster.tsx]: done

File: webapp/src/components/cluster.tsx (hunk 14 above)
Comment (maswin) [reply webapp/src/components/cluster.tsx]: done

File: webapp/src/components/cluster.tsx (hunk 13 above)
Comment (maswin) [reply webapp/src/components/cluster.tsx]: removed

File: webapp/src/components/cluster.tsx (hunk 16 above)
Comment (maswin) [reply webapp/src/components/cluster.tsx]: done

File: webapp/src/components/cluster.tsx (hunk 17 above)
Comment (maswin) [reply webapp/src/components/cluster.tsx]: done

File: webapp/src/components/cluster.tsx (hunk 19 above)
Comment (maswin) [reply webapp/src/components/cluster.tsx]: done

File: webapp/src/components/cluster.tsx (hunk 18 above)
Comment (maswin) [reply webapp/src/components/cluster.tsx]: done

**Comment by maswin** [comment]
> @maswin can you address comments and work with @ebyhr and @Chaho12 ? I would like to get this feature in.

Sure. Addressed all review comments. Hoping to get this feature in.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java (hunk 12 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:236]: fixed

**Comment by Chaho12** [comment]
Build is failing. Please try building locally and fix the failure.

**Comment by maswin** [comment]
> Build is failing. Please try building locally and fix the failure.

Done.

**Comment by maswin** [comment]
@vishalya @Chaho12 - Is this good to be merged?

**Comment by Chaho12** [comment]
@coderabbitai full review

File: webapp/src/components/cluster.tsx [hunk 20]
```diff
@@ -207,6 +219,13 @@ export function Cluster() {
             ]}
             initValue={form?.externalUrl}
           />
+          <Form.TagInput
+            field="tags"
+            label="Tags"
+            initValue={form?.tags}
+            separator={[' ']}
```
Comment (Chaho12) [inline webapp/src/components/cluster.tsx]: What happens if user pastes `env:prod,test` (tag with comma)?
We should add `,` to separator so UI splits instead of producing invalid tag


```suggestion
            separator={[' ', ',']}
```

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java [hunk 21]
```diff
@@ -86,4 +92,18 @@ public void setRoutingGroup(String routingGroup)
...
+    {
+        return tags;
+    }
+
+    @JsonSetter
+    public void setTags(List<String> tags)
+    {
+        this.tags = requireNonNullElse(tags, ImmutableList.<String>of()).stream()
+                .map(tag -> tag == null ? null : tag.trim())
+                .toList();
```
Comment (Chaho12) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: Stream.toList() allows null so it is harmless via add/update (validation rejects), but JSON round-trip `serializes null` in array. Filter instead.
```suggestion
        this.tags = requireNonNullElse(tags, ImmutableList.<String>of()).stream()
                .filter(tag -> tag != null && !tag.isBlank())
                .map(String::trim)
                .toList();
```

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java [hunk 22]
```diff
@@ -247,11 +253,30 @@ private static List<ProxyBackendConfiguration> upcast(List<GatewayBackend> gatew
...
+    }
+
+    @VisibleForTesting
+    static List<String> tagsFromString(String tags)
+    {
+        if (isNullOrEmpty(tags)) {
+            return List.of();
+        }
+        return Splitter.on(',').trimResults().omitEmptyStrings().splitToList(tags);
+    }
```
Comment (Chaho12) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:278]: What happens if there is duplicate in tags? What about deduplicating it?

Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:278]: Added `.distinct()` and removed duplicate tags

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java (hunk 21 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: Done

File: webapp/src/components/cluster.tsx (hunk 20 above)
Comment (maswin) [reply webapp/src/components/cluster.tsx]: Spaces looks legal in a tag (i.e, `ver: trino 476`). So keeping it to only `,`

File: webapp/src/components/cluster.tsx [hunk 23]
```diff
@@ -207,6 +219,13 @@ export function Cluster() {
...
+            placeholder="Type a tag and press Space or Enter"
```
[the remark this answers is not in the document]
Comment (maswin) [reply webapp/src/components/cluster.tsx]: Done

File: webapp/src/components/cluster.tsx [hunk 24]
```diff
@@ -65,6 +65,17 @@ export function Cluster() {
     );
   }
 
+  const tagsRender = (tags: string[]) => {
+    if (!tags || tags.length === 0) return null;
+    return (
+      <>
+        {tags.map(tag => (
+          <Tag key={tag} style={{ marginRight: 4, marginBottom: 2 }}>{tag}</Tag>
```
[the remark this answers is not in the document]
Comment (maswin) [reply webapp/src/components/cluster.tsx]: done

File: webapp/src/components/cluster.tsx (hunk 20 above)
Comment (maswin) [reply webapp/src/components/cluster.tsx]: Similar to the above comment - https://github.com/trinodb/trino-gateway/pull/947#discussion_r3661639126

File: docs/gateway-api.md [hunk 25]
```diff
@@ -79,7 +83,8 @@ Returns a JSON array of Trino cluster:
         "proxyTo": "http://localhost:8083",
         "active": true,
         "routingGroup": "adhoc",
-        "externalUrl": "http://localhost:8084"
+        "externalUrl": "http://localhost:8084",
+        "tags": []
```
[the remark this answers is not in the document]
Comment (maswin) [reply docs/gateway-api.md:88]: done

File: docs/backend-tags.md [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+the only mechanism for influencing routing decisions is the `routingGroup` field,
+which maps a query to a named cluster group. There is no way to store additional
+metadata on a backend that routing logic could inspect. Tags provide that
+foundation — a freeform list of strings on each backend that can be read by
+custom routing rules, scripts, or future gateway enhancements to enable more
+sophisticated routing behavior.
+
+Tags are stored as a list of strings on the backend and are returned in all
+backend API responses. A common convention is `key:value` pairs (e.g. `env:prod`,
+`team:data-eng`), but any string is accepted.
```
[the remark this answers is not in the document]
Comment (maswin) [reply docs/backend-tags.md]: done

**Comment by maswin** [comment]
@Chaho12 - Is this good to be merged?

File: gateway-ha/src/main/resources/postgresql/V5__add_tags_to_gateway_backend.sql [hunk 27]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE gateway_backend
+ADD COLUMN tags TEXT NULL;
```
Comment (Chaho12) [inline gateway-ha/src/main/resources/postgresql/V5__add_tags_to_gateway_backend.sql:2]: Shouldn't this be IF NOT exists? Because flyway calls this and `tags` already exist at `gateway-ha/src/main/resources/gateway-ha-persistence-mysql.sql`

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java [hunk 28]
```diff
@@ -229,6 +231,10 @@ private static void validateBackendConfiguration(ProxyBackendConfiguration backe
         checkArgument(backend.getProxyTo() != null, "Backend proxyTo URL cannot be null");
         checkArgument(backend.getRoutingGroup() != null, "Backend routing group cannot be null");
         checkArgument(backend.getExternalUrl() != null, "Backend external url cannot be null");
+        for (String tag : backend.getTags()) {
+            checkArgument(tag != null && !tag.isBlank(), "Backend tag cannot be null or blank");
```
Comment (Chaho12) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:235]: Agreed on the comma check at line 236 — that one must stay. 
My comment was about line 235 (`tag != null && !tag.isBlank()`), which is unreachable.
`setTags` already filters null and blank entries, and `testAddBackendNormalizesTags` asserts exactly that (`getTags()` returns `["env:prod", "size:XL"]` for an input containing `null` and `"   "`). 

No test asserts the "cannot be null or blank" message, only the comma one.

There is also a policy mismatch worth settling: docs/backend-tags.md says blank tags are **ignored**, while line 235 claims they are **rejected**.

File: docs/backend-tags.md [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+        "proxyTo": "http://localhost:8083",
+        "active": true,
+        "routingGroup": "adhoc",
+        "tags": ["env:prod", "team:data-eng", "size:XL"]
+    }'
+```
+
+!!! note "Commas are not allowed in tag values"
+    A tag value cannot contain a `,` character — the gateway will reject the
+    request with an `IllegalArgumentException`. If you need to express
```
Comment (Chaho12) [inline docs/backend-tags.md:35]: Are you sure that this is IllegalArgumentException? and not http 500 error?

**Comment by Chaho12** [comment]
Also please rebase, there seems conflicts

**Comment by maswin** [comment]
> Also please rebase, there seems conflicts

Done

File: docs/backend-tags.md (hunk 29 above)
Comment (maswin) [reply docs/backend-tags.md]: Fixed. Dropped the exception name from the docs and added generic message

File: gateway-ha/src/main/resources/postgresql/V5__add_tags_to_gateway_backend.sql (hunk 27 above)
Comment (maswin) [reply gateway-ha/src/main/resources/postgresql/V5__add_tags_to_gateway_backend.sql:2]: Fixed by removing tags from both gateway-ha-persistence-{mysql,postgres}.sql rather than making V5 idempotent - ADD COLUMN IF NOT EXISTS isn't portable (MySQL and Oracle don't support it), and those scripts are the pre-Flyway baseline, which is why V2/V3's routing_group/external_url aren't in them either.
HaGatewayTestUtils.seedRequiredData now applies the V5 script after the bootstrap script, since the H2 unit-test path never runs Flyway.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java (hunk 28 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java]: Fixed

**Comment by Chaho12** [comment]
PTAL @vishalya

File: docs/backend-tags.md
Comment (Chaho12) [inline docs/backend-tags.md:1]: I am not so sure where to put, but maybe at release note and backend tags?
We should mention that if `runMigrationsEnabled` is false, and tags column is not added manually (even if you don't use tag) creation/edit backend fails. 

```md
!!! warning "Manual migrations"
    If `runMigrationsEnabled` is `false`, apply
    `V5__add_tags_to_gateway_backend.sql` before upgrading. All backend
    create/update requests fail until the `tags` column exists, even when no
    tags are used.
```

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java [hunk 30]
```diff
@@ -229,6 +231,9 @@ private static void validateBackendConfiguration(ProxyBackendConfiguration backe
         checkArgument(backend.getProxyTo() != null, "Backend proxyTo URL cannot be null");
         checkArgument(backend.getRoutingGroup() != null, "Backend routing group cannot be null");
         checkArgument(backend.getExternalUrl() != null, "Backend external url cannot be null");
+        for (String tag : backend.getTags()) {
```
Comment (Chaho12) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:234]: Should we have limitations on number of tags? Currently it is infinite, but all this loads to memory

File: gateway-ha/src/main/resources/oracle/V5__add_tags_to_gateway_backend.sql [hunk 31]
```diff
@@ -0,0 +1 @@
+ALTER TABLE gateway_backend ADD (tags CLOB NULL);
```
Comment (Chaho12) [inline gateway-ha/src/main/resources/oracle/V5__add_tags_to_gateway_backend.sql]: Use varchar  instead of clob. No need for this. 
```suggestion
ALTER TABLE gateway_backend ADD (tags VARCHAR2(4000) NULL);
```

Image

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java [hunk 32]
```diff
@@ -86,4 +93,20 @@ public void setRoutingGroup(String routingGroup)
...
+        return tags;
+    }
+
+    @JsonSetter
+    public void setTags(List<String> tags)
+    {
+        this.tags = requireNonNullElse(tags, ImmutableList.<String>of()).stream()
+                .filter(tag -> tag != null && !tag.isBlank())
+                .map(String::trim)
+                .distinct()
```
Comment (Chaho12) [inline gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: Instead of blank -> trim, isBlank/trim is checked twice. Why don't we do trim first. 
```suggestion
                .filter(Objects::nonNull)
                .map(String::trim)
                .filter(tag -> !tag.isEmpty())
```

**Review by Chaho12 — APPROVED** [review]
LGTM 👍 
Just minor fixes :)

File: docs/backend-tags.md
Comment (maswin) [reply docs/backend-tags.md:1]: I have added it to tags documentation as recommended. Will edit the PR description to put it in release notes as well

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java (hunk 30 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/HaGatewayManager.java:234]: Feels like a pre mature optimization, would recommend leaving it to the user on the limits

File: gateway-ha/src/main/resources/oracle/V5__add_tags_to_gateway_backend.sql (hunk 31 above)
Comment (maswin) [reply gateway-ha/src/main/resources/oracle/V5__add_tags_to_gateway_backend.sql]: done

File: gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java (hunk 32 above)
Comment (maswin) [reply gateway-ha/src/main/java/io/trino/gateway/ha/config/ProxyBackendConfiguration.java]: done
